implementation/src/main/java/io/smallrye/jwt/auth/AbstractDpopTokenValidator.java

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright 2020 Red Hat, Inc, and individual contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.smallrye.jwt.auth;
+
+import static org.jose4j.jwa.AlgorithmConstraints.ConstraintType.PERMIT;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Map;
+
+import org.eclipse.microprofile.jwt.Claims;
+import org.eclipse.microprofile.jwt.JsonWebToken;
+import org.jose4j.jwa.AlgorithmConstraints;
+import org.jose4j.jwt.JwtClaims;
+import org.jose4j.jwt.consumer.InvalidJwtException;
+import org.jose4j.jwt.consumer.JwtConsumer;
+import org.jose4j.jwt.consumer.JwtConsumerBuilder;
+import org.jose4j.jwt.consumer.Validator;
+import org.jose4j.jwx.JsonWebStructure;
+import org.jose4j.keys.resolvers.EmbeddedJwkVerificationKeyResolver;
+import org.jose4j.lang.HashUtil;
+import org.jose4j.lang.JoseException;
+
+import io.smallrye.jwt.auth.principal.JWTAuthContextInfo;
+import io.smallrye.jwt.auth.principal.ParseException;
+
+/**
+ * Common functionality for classes implementing DPoP token proofing using
+ * HTTP request headers and a valid DPoP-bound {@link JsonWebToken}.
+ *
+ * @author Aaron Coburn {@literal <acoburn@apache.org>}
+ */
+public abstract class AbstractDpopTokenValidator {
+
+    protected static final String DPOP_HEADER = "DPoP";
+    protected static final String DPOP_HTTP_URI_CLAIM = "htu";
+    protected static final String DPOP_HTTP_METHOD_CLAIM = "htm";
+    protected static final String DPOP_JWT_TYPE = "dpop+jwt";
+    protected static final String JSON_WEB_KEY_THUMBPRINT = "jkt";
+
+    private final JWTAuthContextInfo authContextInfo;
+
+    protected AbstractDpopTokenValidator(JWTAuthContextInfo authContextInfo) {
+        this.authContextInfo = authContextInfo;
+    }
+
+    public void verify(JsonWebToken accessToken) throws ParseException {
+        final String dpop = getDpopHeaderValue();
+        final String thumbprint = getDpopKeyThumbprint(accessToken);
+        // DPoP validation is only relevant for access tokens with a bound confirmation key.
+        if (thumbprint != null) {
+            if (dpop != null) {
+                final JwtConsumer parser = new JwtConsumerBuilder()
+                        .setRequireJwtId()
+                        .setExpectedType(true, DPOP_JWT_TYPE)
+                        .setJwsAlgorithmConstraints(new AlgorithmConstraints(PERMIT,
```

Comment:
This algorithm is used for decrypting the claims or inner signed tokens when dealing with the access token. I believe you meant `getSignatureAlgorithm`.
But you have a good point about `ES256`/etc - I propose that we have a new property added, `smallrye.jwt.verify.dpop.signature-algorithm`, its default value is the same as that of `smallrye.jwt.verify.signature-algorithm` which is `RS256` but the users can choose `ES256`

Reply:
Yes, I did mean `getSignatureAlgorithm`. I like your suggestion for adding a `smallrye.jwt.verify.dpop.signature-algorithm` config property